### PR TITLE
Refactor MYOB

### PR DESCRIPTION
**Pull request by thukim**
Related card: https://trello.com/c/hjhtIaQo/582-chore-turn-timeout-expected-exceptions-to-notifications-in-3-payroll-gems-to-help-clean-up-airbrake

- If MYOB couldn’t create payroll_detail object, store that error into  EH member
- Remove unnecessary logging info
- Split payslips into groups of 20 items each to bulk import to EH
- No need to exclude any leave balances if members have no existing leave balances in EH
- Raise errors when GET request to EH API fails

### CODE REVIEW

File: lib/eh/myob/sync/leave_request.rb [hunk 1]
```diff
@@ -8,13 +8,18 @@ class LeaveRequest
...
         rescue OAuth2::Error => e
-          errors = e.response.parsed['Errors'].map {|e| e['Message'] }
           Eh::Myob.config.logger.info({
                             exception: e,
                             parsed_error: e.response.parsed['Errors'],
                             method: 'LeaveRequest#push',
-                            attributes: { eh_context: @eh_context, myob_context: @myob_context }
+                            attributes: { eh_context: @eh_context }
                           })
+          errors = e.response.parsed['Errors']
```
Comment (mmeeoorroo) [inline lib/eh/myob/sync/leave_request.rb]: According to `oauth2` source, `response.parsed` wouldn't be a `Hash` https://github.com/oauth-xx/oauth2/blob/v1.3.0/lib/oauth2/error.rb#L11. So this line would raise `NoMethod` error

File: lib/eh/myob/sync/employee_file.rb [hunk 2]
```diff
@@ -254,7 +254,7 @@ def exclude_balances_not_exist_in_eh(formatted_members, eh_context)
                 myob_leave.delete(:leave_category_external_id)
               end
               member[:leave_balances_attributes].select! { |lb| lb[:leave_category_id] }
-            end
+            end if members
```
Comment (mmeeoorroo) [inline lib/eh/myob/sync/employee_file.rb]: Is this necessary? The `formatted_members.select` would return an empty `members` so `members.each` won't execute. Btw, an empty array is a non-nil object, so this `if` is useless. I tested in console:

```
2.3.3 :013 >   arr = [1,2].select { |i| i == 0 }
 => []
2.3.3 :014 > puts 'hello' if arr
hello
 => nil
```

Comment (thukim) [reply lib/eh/myob/sync/employee_file.rb]: I intended to fix the old airbrake bug, which I fixed it before, by using the `formatted_members.select` above. thats why this is unnecessary, will remove it later, tks!

File: spec/eh/myob/sync/payslip_spec.rb [hunk 3]
```diff
@@ -10,28 +10,89 @@
...
+            {"first_name"=>"Bi", "last_name"=>"ghe", "name"=>"Bi ghe", "organisation_id"=>8483, "payrun_attributes"=>{"payment_date"=>"2015-09-02T00:00:00", "period_start_date"=>"2015-08-27T00:00:00", "period_end_date"=>"2015-09-02T00:00:00"}},
+            {"first_name"=>"Bi", "last_name"=>"ghe", "name"=>"Bi ghe", "organisation_id"=>8483, "payrun_attributes"=>{"payment_date"=>"2015-09-02T00:00:00", "period_start_date"=>"2015-08-27T00:00:00", "period_end_date"=>"2015-09-02T00:00:00"}},
+            {"first_name"=>"Bi", "last_name"=>"ghe", "name"=>"Bi ghe", "organisation_id"=>8483, "payrun_attributes"=>{"payment_date"=>"2015-09-02T00:00:00", "period_start_date"=>"2015-08-27T00:00:00", "period_end_date"=>"2015-09-02T00:00:00"}},
+            {"first_name"=>"Bi", "last_name"=>"ghe", "name"=>"Bi ghe", "organisation_id"=>8483, "payrun_attributes"=>{"payment_date"=>"2015-09-02T00:00:00", "period_start_date"=>"2015-08-27T00:00:00", "period_end_date"=>"2015-09-02T00:00:00"}},
```
Comment (khoido) [inline spec/eh/myob/sync/payslip_spec.rb]: Why do you need to create a lot of duplicated rows ? Can we move them to VCR file ?

File: lib/eh/myob/sync/base.rb [hunk 4]
```diff
@@ -6,7 +6,7 @@ def pull(eh_context:, myob_context:, all: false)
           method, formatted_data = prepare_pull_data(myob_context, true)
           errors = if all
             conflict_items = formatted_data.present? ? transform_with_eh_ids(eh_context, formatted_data) : []
-            return conflict_items[:sync_errors] if conflict_items.instance_of?(Hash)
+            return conflict_items if conflict_items.instance_of?(Hash)
```
Comment (mmeeoorroo) [inline lib/eh/myob/sync/base.rb]: Why do you change to return a hash instead of an array? Did you also update the caller of this method?

Comment (thukim) [reply lib/eh/myob/sync/base.rb]: yes, this change is to fix a bug, the caller expects this to return a hash instead.

File: spec/eh/myob/sync/payslip_spec.rb (hunk 3 above)
Comment (thukim) [reply spec/eh/myob/sync/payslip_spec.rb]: coz I need to test for a group of more than 20 payslips, to see how many requests it make to the EH API :P

File: lib/eh/myob/sync/leave_request.rb (hunk 1 above)
Comment (thukim) [reply lib/eh/myob/sync/leave_request.rb]: eg:
```
 exception=#<OAuth2::Error: : 
{
  "Errors": [
    {
      "Name": "CompanyFilePreferenceForTimesheet",
      "Message": "Company File preference to 'Use Timesheet for...' should be ON",
      "AdditionalDetails": null,
      "ErrorCode": 34003,
      "Severity": "Error"
    }
  ],
```

**Pull request merged**